lib/spack/spack/binary_distribution.py

```diff
@@ -0,0 +1,260 @@
+        os.utime(path1, None)       
+    path2='%s.asc' % spec_file
+    with open(path2,'a'):
+        os.utime(path2, None)       
+           
+
+    spackfile_path = os.path.join(outdir, tarball_path_name(spec, '.spack'))
+    with tarfile.open(spackfile_path, 'w') as tar:
+        tar.add(name = '%s' % tarfile_path, arcname = '%s' % tarfile_name)
+        tar.add(name = '%s' % spec_file, arcname = 'spec.yaml')
+        tar.add(name = '%s.asc' % tarfile_path , arcname = '%s.asc' % tarfile_name )
+        tar.add(name = '%s.asc' % spec_file, arcname = 'spec.yaml.asc' ) 
+        os.remove(tarfile_path)
+        os.remove(path1)
+        os.remove(path2)
+    tty.msg(tarfile_path)
+
+
+def download_tarball(package):
+    """
+    Download binary tarball for given package into stage area
+    Return True if successful
+    """
+    mirrors = spack.config.get_config('mirrors')
+    if len(mirrors) == 0:
+        tty.die("Please add a spack mirror to allow " +
+                "download of pre-compiled packages.")
+    tarball = tarball_path_name(package.spec, '.spack')
+    for key in mirrors:
+        url = mirrors[key] + "/" + tarball
+        # print url
+        # stage the tarball into standard place
+        stage = Stage(url, name=package.stage.path)
+        try:
+            stage.fetch()
+            return True
+        except fs.FetchError:
+            next
+    return False
+
+
+def extract_tarball(package):
+    """
+    extract binary tarball for given package into install area
+    """
+    tarball = tarball_name(package.spec,'.spack')
+    local_tarball = package.stage.path + "/" + tarball
+    mkdirp(package.prefix)
+    tarfile_name = tarball_name(package.spec, '.tar.gz')
+    tarfile_path = os.path.join(package.stage.path, tarfile_name)
+    with tarfile.open(local_tarball, 'r') as tar:
+        tar.extract('spec.yaml', package.stage.path)
+        tar.extract('spec.yaml.asc', package.stage.path)
+
+        # spack gpg verify os.path.join(package.prefix, 'spec.yaml')
+
+        tar.extract(tarfile_name, package.stage.path)
+        tar.extract(tarfile_name + '.asc', package.stage.path)
+
+        # spack gpg verify tarfile_path
+
+    with tarfile.open(tarfile_path, 'r') as tar:
+        tar.extractall(path = os.path.dirname(package.prefix))
+
+    #os.remove(tarfile_path)
+    #os.remove(tarfile_path + '.asc')
+
+
+def relocate_package(package):
+    """
+    Relocate the given package
+    """
+    buildinfo = read_buildinfo_file(package)
+    new_path = spack.store.layout.root
+    old_path = buildinfo['buildpath']
+    if old_path == new_path:
+        return True  # No need to relocate
+
+    tty.warn("Using experimental feature for relocating package from",
+             "%s to %s." % (old_path, new_path))
+
+    # as we may need patchelf, find out where it is
+    patchelf_executable = ''
+    if platform.system() != 'Darwin':
+        patchelf_spec = spack.cmd.parse_specs("patchelf", concretize=True)[0]
```

Comment:
It also seems that the cmd infrastructure just swallows exceptions? I see the exceptions in the test, but the command just silently does nothing :/ . `tty` also doesn't actually output anything.